### DISCUSSION TITLE
How to Change Tailwind Styles Based On React Variable

### DISCUSSION THREAD

**Discussion by JimLynchCodes** [tailwindlabs/tailwindcss#1507] [General]
Hello! I have a header bar, and I'm trying to have it display / hide when the user click a button a toggles a boolean React component state variable. 

This is how I would like it to look when expanded:
![Screen Shot 2020-03-29 at 11 49 27 AM](https://user-images.githubusercontent.com/5354163/77853674-696c2080-71b3-11ea-98e7-c5e04d2a5417.png)

Here is my code for the toggle button:
```
 <div className="my-8 block lg:hidden" onClick={() => toggleOpen(!expanded)}>
          <button className="flex items-center px-3 py-2 border rounded text-teal-200 border-teal-400 hover:text-white hover:border-white">
            <svg className="fill-current h-3 w-3" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><title>Menu</title><path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z" /></svg>
          </button>
        </div>
```

Here is a snippet of the div that I would like to hide:
```
 <div className={"w-full block flex-grow lg:flex lg:items-center lg:w-auto"}>
```
 
When I add the "hidden" tailwind class (short for `display: hidden;`) then the expandable part does hide correctly:

```
<div className={"w-full block flex-grow lg:flex lg:items-center lg:w-auto hidden"}>
```

![Screen Shot 2020-03-29 at 11 51 50 AM](https://user-images.githubusercontent.com/5354163/77853729-b05a1600-71b3-11ea-9901-2e869fff8fef.png)

If I try to modify the classes dynamically based on the React variable though... it doesn't work.

```
 {JSON.stringify(expanded)}

 <div className={"w-full flex-grow lg:flex lg:items-center lg:w-auto " + expanded ? 'block' : 'hidden'}>
```

**Comment by JimLynchCodes** [comment]
![Screen-Recording-2020-03-29-at-1](https://user-images.githubusercontent.com/5354163/77853869-9240e580-71b4-11ea-8470-52c033c6e00b.gif)

**Reply by venus0311** [reply]
thk

**Comment by PILO7980** [comment]
```<div className={"w-full flex-grow lg:flex lg:items-center lg:w-auto " + (expanded ? 'block' : 'hidden')}>```

**Reply by tolumide-ng** [reply]
You could also modify this, to make it more readable. Something like => 
```
const styles = {divClass: 'w-full flex-grow lg:flex lg:items-center lg:w-auto'}
```
...and on the element, you can have something like this => 
```
<div className={`${styles.divClasses} ${expanded ? 'block' : 'hidden'}`} >whatever</div>
```

Longer but cleaner ...at least in my opinion

**Reply by cschroeter** [reply]
> `<div className={"w-full flex-grow lg:flex lg:items-center lg:w-auto " + (expanded ? 'block' : 'hidden')}>`

oh boi is this hard to ready

**Reply by theabhayprajapati** [reply]
> you can use 
> 
> ```
> {`text-white ${true?'text-green-500' : 'text-red-500'`}
> ```
> 
> with curly braces and backticks you can do easily.

**Reply by theabhayprajapati** [reply]
> > `<div className={"w-full flex-grow lg:flex lg:items-center lg:w-auto " + (expanded ? 'block' : 'hidden')}>`
> 
> oh boi is this hard to ready
> 

✅ Better one.
> you can use 
> 
> ```
> {`text-white ${true?'text-green-500' : 'text-red-500'`}
> ```
> 
> with curly braces and backticks you can do easily.

**Reply by sriram-balaganeshan** [reply]
this is i used in my project and its works perfect  
  className={
                    " flex font-sans font-medium text-center text-xs rounded !h-[40px] m-[8px] group hover:bg-black-opacity-15 " +
                    `${val.active ? "bg-activeSidebar" : ""} `
                  }
need space for + , ` these two

**Reply by mujtabafromthedarkside** [reply]
It took me an hour to realize those are backticks on the outside

**Reply by ginogardiola** [reply]
react is garbage

**Comment by JimLynchCodes** [comment]
Thanks @PILO7980 !

I would think these are equivalent. Why does the it work only when there are extra parentheses? 🤔

**Reply by adamwathan** [reply]
I'm guessing that without the parentheses it's being interpreted like this:

```
 <div className={("w-full flex-grow lg:flex lg:items-center lg:w-auto " + expanded) ? 'block' : 'hidden'}>
```

**Comment by robertt** [comment]
This works, but I really hate the syntax - I know this has nothing to do with tailwind but I hate having to write stuff like that. I really wish there was some :class like there there is in alpine and vue, making it a lot more readable. 

(I'm also not sure if it's only me but sometimes I have to put my conditional classes at the end instead of at the start, not sure if this is me just being bad at tailwind but it looks even worse when you have a combination).

**Reply by adamwathan** [reply]
Use this package 👍 

https://github.com/JedWatson/classnames#readme

**Reply by brielov** [reply]
@adamwathan Does `classnames` work with `purgecss`?

**Reply by peacepostman** [reply]
I'm using className but classes are not included in the final css, maybe i'm missing something with my configuration

**Reply by peacepostman** [reply]
Ow boy was i wrong, i was missing the configuration hint for disabled:opacity. So yes, it works great with `classNames`

**Reply by dillonstreator** [reply]
https://github.com/lukeed/clsx is likely a better option for it's speed and footprint and it can act as a drop-in replacement for classnames

**Reply by Axedyson** [reply]
Do classNames work with [tailwindcss vscode extension](https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss)

**Reply by amannn** [reply]
I'd be curious if there's a variant of a utility like `clsx` that handles overrides? E.g.:

```jsx
clsx('p-4', 'p-2'); // p-2
clsx('p-4', 'pt-2'); // pt-2 px-4 pb-4
```

That would be really handy to allow for overrides in components. Often you'd probably want to avoid this, but I think it can be a nice escape hatch and such a utility function would support the pattern without specificity issues.

**Reply by Pika-Pool** [reply]
@peacepostman , how did you make tailwind work with classnames? I'm using tailwind 3.0, and when I do this:
```
className={classNames(
	'flex justify-center items-center gap-2 rounded lg:px-8 lg:py-3 border border-primary',
	{ 'color-primary bg-transparent': votedByMe },
)}
```
the conditional classes are not included in the generated css.

**Comment by bartenra** [comment]
You can also use template strings:

```jsx
export const TextInput = ({error, ...props}: TextInputProps) => {
  return (
    <div className="flex flex-row justify-center">
      <input type="text" {...props} className={`w-full p-3 my-3 text-xl border-4 ${error ? "border-red-400" : ""}`} />
    </div>
  )
}
```

**Reply by davidparys** [reply]
How do you handle the ugly empty spaces for falsy values?

**Reply by theolvq** [reply]
You could use the short-circuit operator
````javascript
export const TextInput = ({error, ...props}: TextInputProps) => {
  return (
    <div className="flex flex-row justify-center">
      <input type="text" {...props} className={`w-full p-3 my-3 text-xl border-4 ${error &&  "border-red-400" }`} />
    </div>
  )
}
````

**Reply by Axedyson** [reply]
Okay but if `error` is `false` this is what gets rendered: 
```TypeScript
`w-full p-3 my-3 text-xl border-4 `
```
As you can see there is still a whitespace at the end of the string 😕

**Reply by tonny008** [reply]
> 
short-circuit operator `${error &&  "border-red-400" }`. if error = false, the output is false

**Reply by davidparys** [reply]
> > 
> short-circuit operator `${error &&  "border-red-400" }`. if error = false, the output is false
> 

Yes sure but since it's a template literal, if `error` it will actually return the error inside the class

**Comment by matheuslemke** [comment]
Just to complement with another case... I've tried to do something like `bg-${color}`, but **DIDN'T WORK**:
```jsx
// color being like red-200
const Area = ({ title, color }: AreaProps) => {
  return (
    <div className={`h-[50vh] flex items-center justify-center font-sans bg-${color}`}>
      <div>{title}</div>
    </div>
  )
}
```

You had to provide the entire property, like:
```jsx
// bgColor is something like bg-red-200
const Area = ({ title, bgColor }: AreaProps) => {
  return (
    <div className={`h-[50vh] flex items-center justify-center font-sans ${bgColor}`}>
      <div>{title}</div>
    </div>
  )
}
```

**Reply by rutvik24** [reply]
Thanks bro

**Reply by DDSAO** [reply]
I have encountered the same issue. Can somebody provide some insights about how and why?

**Reply by wongjn** [reply]
From [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):
> The most important implication of how Tailwind extracts class names is that it will only find classes that exist *as complete unbroken strings* in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS:

**Reply by rutvik24** [reply]
Instead of passing variable to the class make variable with class then pass it it will work
Like const hClass = h-[2px];
className=`${hClass}`

Try above one instead of this
const height = 2px;
className=`h-${height}`

**Reply by stub4ever** [reply]
Thank you bro

**Comment by Axedyson** [comment]
Simple solution just use this [tool](https://github.com/netlify/classnames-template-literals)

**Comment by yousif-khalil-abdulkarim** [comment]
You could use the [class variance authority](https://github.com/joe-bell/cva) library which makes it easy to create component libraries with tailwind.

**Comment by kubo550** [comment]
Maybe you can achieve what you need using the `style` property where you can use variables.

```javascript
        <img
              className="absolute bottom-0"
              style={{ transform: `rotate(${needleRotation}deg)` }}
              src="image.png"
            />
```

**Comment by estebangarviso** [comment]
Try to use prop-types https://blog.logrocket.com/theming-react-components-tailwind-css/

```js
// src/components/Button.js
import themeProps from "./themeProps";
const Button = ({ children, color, ...rest }) => {
  return (
    <button className={`rounded-md bg-${color} text-text-base px-3 py-1`} {...rest}>
      {children}
    </button>
  );
};

Button.propTypes = {
  ...themeProps,
};

Button.defaultProps = {
  color: "primary",
};

export default Button;
```

**Comment by manuelhonoredesousa** [comment]
I hope you see this -> https://tailwindcss.com/blog/tailwindcss-v3-2#data-attribute-variants
